lit_stablelm/model.py

```diff
@@ -0,0 +1,220 @@
+
+    def forward(self, idx: torch.Tensor) -> torch.Tensor:
+        _, t = idx.size()
+        assert (
+            t <= self.config.block_size
+        ), f"Cannot forward sequence of length {t}, block size is only {self.config.block_size}"
+
+        # forward the LLaMA model itself
+        x = self.transformer.wte(idx)  # token embeddings of shape (b, t, n_embd)
+
+        for block in self.transformer.h:
+            x = block(x)
+        x = self.transformer.ln_f(x)
+
+        logits = self.lm_head(x)  # (b, t, vocab_size)
+
+        return logits
+
+    @classmethod
+    def from_name(cls, name: str) -> Self:
+        return cls(StableLMConfig.from_name(name))
+
+
+class Block(nn.Module):
+    def __init__(self, config: StableLMConfig) -> None:
+        super().__init__()
+        self.norm_1 = nn.LayerNorm(config.n_embd)
+        self.attn = CausalSelfAttention(config)
+        self.norm_2 = nn.LayerNorm(config.n_embd)
+        self.mlp = MLP(config)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        # gpt-j residual
+        x = x + self.attn(self.norm_1(x)) + self.mlp(self.norm_2(x))
+        return x
+
+
+class CausalSelfAttention(nn.Module):
+    def __init__(self, config: StableLMConfig) -> None:
+        super().__init__()
+        assert config.n_embd % config.n_head == 0
+
+        # key, query, value projections for all heads, but in a batch
+        self.attn = nn.Linear(config.n_embd, 3 * config.n_embd, bias=True)
+        # output projection
+        self.proj = nn.Linear(config.n_embd, config.n_embd, bias=True)
+
+        self.n_head = config.n_head
+        self.n_embd = config.n_embd
+        self.block_size = config.block_size
+        self.rotary_percentage = config.rotary_percentage
+        self.rope_cache: Optional[Tuple[torch.Tensor, torch.Tensor]] = None
+
+        if self.rotary_percentage != 1.0:
+            self.register_buffer(
+                "bias",
+                torch.tril(torch.ones(config.block_size, config.block_size)).view(
+                    1, 1, config.block_size, config.block_size
+                ),
+            )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        B, T, C = x.size()  # batch size, sequence length, embedding dimensionality (n_embd)
+
+        qkv = self.attn(x)
+        head_size = C // self.n_head
+        qkv = qkv.view(B, T, self.n_head, 3 * head_size).transpose(1, 2)
+        q, k, v = qkv.split(head_size, dim=-1)  # (B, nh, T, hs)
+
+        n_elem = int(self.rotary_percentage * head_size)
+        if self.rope_cache is None:
+            self.rope_cache = build_rope_cache(self.block_size, n_elem, x.dtype, x.device)
+        cos, sin = self.rope_cache
+        cos, sin = cos[:T], sin[:T]
+
+        q_roped = apply_rope(q[..., :n_elem], cos, sin)
+        k_roped = apply_rope(k[..., :n_elem], cos, sin)
+        q = torch.cat((q_roped, q[..., n_elem:]), dim=-1)
+        k = torch.cat((k_roped, k[..., n_elem:]), dim=-1)
```

Comment:
key change: the entire rope implementation is changed to support rotary percentages